python/mlir_structured/dialects/indexing.py

```diff
@@ -29,90 +46,199 @@
+
+  def __call__(cls, *args, **kwargs):
+    if len(args) != 1:
+      raise ValueError("Only one non-kw arg supported.")
+    arg = args[0]
+    if isinstance(arg, (OpView, Operation, Value)):
+      if isinstance(arg, (Operation, OpView)):
+        assert len(arg.results) == 1
+      val = get_op_result_or_value(arg)
+      arg_copy = None
+    elif isinstance(arg, (int, float, bool, np.ndarray)):
+      dtype = kwargs.get("dtype")
+      index = kwargs.get("index")
+      arg_copy = deepcopy(arg)
+      val = constant(arg, dtype, index).result
+    else:
+      raise NotImplementedError(
+          f"{cls.__name__} doesn't support wrapping {arg}.")
+
+    cls_obj = cls.__new__(cls, val)
+    cls.__init__(cls_obj, val, arg_copy)
+    return cls_obj
+
+
+class ArithValue(metaclass=ArithValueMeta):
+
+  def __new__(cls, val):
+    # The call chain after this is the __init__ here *and then* Value.__init__
+    return super().__new__(cls, val)
+
+  def __init__(self,
+               val,
+               arg: Optional[Union[int, float, bool, np.ndarray]] = None):
+    self.__arg = arg
+    super().__init__(val)
+
+  @lru_cache(maxsize=1)
+  def is_constant(self) -> bool:
+    return isinstance(self.owner.opview, arith.ConstantOp)
+
+  @lru_cache(maxsize=1)
   def __str__(self):
     s = res_val_reg.findall(super().__str__())
     assert len(s) == 1
-    return f"Scalar({s[0]}, {self.type})"
+    if self.is_constant():
+      v = str(self.literal_value)
+      v = f"{v[:10]}..." if len(v) > 10 else v
+      return f"{self.__class__.__name__}({s[0]}, {self.type}, {v})"
+    else:
+      return f"{self.__class__.__name__}({s[0]}, {self.type})"
+
+  @lru_cache(maxsize=1)
+  def __repr__(self):
+    return str(self)
+
+  def __binary_op(self, other: "Tensor", op: str) -> "Tensor":
+    assert op in {"add", "sub", "mul"}
+    if self.type != other.type:
+      raise ValueError(f"{self=} {other=} must have the same type.")
+
+    if self.is_constant() and other.is_constant():
+      lhs, rhs = self.literal_value, other.literal_value
+      op = operator.attrgetter(op)(operator)
+    else:
+      op = op.capitalize()
+      lhs, rhs = self, other
+      if _is_floating_point_type(self.dtype):
+        op = getattr(arith, f"{op}FOp")
+      elif _is_integer_like_type(self.dtype):
+        op = getattr(arith, f"{op}IOp")
+      else:
+        raise NotImplementedError(f"Unsupported '{op}' operands: {lhs}, {rhs}")
+    return self.__class__(op(lhs, rhs))
 
-  def __add__(self, other) -> "Scalar":
-    return Scalar(_body_builder._binary_add(self, other))
+  __add__ = partialmethod(__binary_op, op="add")
+  __sub__ = partialmethod(__binary_op, op="sub")
+  __mul__ = partialmethod(__binary_op, op="mul")
 
-  def __sub__(self, other) -> "Scalar":
-    return Scalar(_body_builder._binary_sub(self, other))
 
-  def __mul__(self, other) -> "Scalar":
-    return Scalar(_body_builder._binary_mul(self, other))
+class Scalar(ArithValue, ScalarValue):
```

Comment:
Order here matters - determines [method resolution order](https://gist.github.com/igormolybog/4487d8bf236005fc0ebefa9160c91ba8) - e.g., you want the `__new__`, `__init__`, and `__str__` from `ArithValue` not `ScalarValue`.